sycl/unittests/scheduler/InOrderQueueHostTaskDeps.cpp

```diff
@@ -0,0 +1,85 @@
+//==-------- InOrderQueueHostTaskDeps.cpp --- Scheduler unit tests ---------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "SchedulerTest.hpp"
+#include "SchedulerTestUtils.hpp"
+
+#include <helpers/CommonRedefinitions.hpp>
+#include <helpers/PiImage.hpp>
+#include <helpers/PiMock.hpp>
+
+#include <detail/event_impl.hpp>
+#include <detail/queue_impl.hpp>
+
+#include <gtest/gtest.h>
+
+class TestKernel;
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+namespace detail {
+template <> struct KernelInfo<TestKernel> {
+  static constexpr unsigned getNumParams() { return 0; }
+  static const kernel_param_desc_t &getParamDesc(int) {
+    static kernel_param_desc_t Dummy;
+    return Dummy;
+  }
+  static constexpr const char *getName() { return "TestKernel"; }
+  static constexpr bool isESIMD() { return false; }
+  static constexpr bool callsThisItem() { return false; }
+  static constexpr bool callsAnyThisFreeFunction() { return false; }
+};
+
+} // namespace detail
+} // namespace sycl
+} // __SYCL_INLINE_NAMESPACE(cl)
+
+using namespace sycl;
+
+size_t GEventsWaitCounter = 0;
+
+inline pi_result redefinedEventsWait(pi_uint32 num_events,
+                                     const pi_event *event_list) {
+  if (num_events > 0) {
+    GEventsWaitCounter++;
+  }
+  return PI_SUCCESS;
+}
+
+TEST_F(SchedulerTest, InOrderQueueHostTaskDeps) {
+  default_selector Selector;
+  platform Plt{default_selector()};
+  if (Plt.is_host()) {
+    std::cout << "Not run due to host-only environment\n";
+    return;
+  }
+
+  unittest::PiMock Mock{Plt};
+  setupDefaultMockAPIs(Mock);
+  Mock.redefine<detail::PiApiKind::piEventsWait>(redefinedEventsWait);
+
+  context Ctx{Plt};
+  queue InOrderQueue{Ctx, Selector, property::queue::in_order()};
+
+  kernel_bundle KernelBundle =
+      sycl::get_kernel_bundle<sycl::bundle_state::input>(Ctx);
+  auto ExecBundle = sycl::build(KernelBundle);
+
+  event Evt = InOrderQueue.submit([&](sycl::handler &CGH) {
+    CGH.use_kernel_bundle(ExecBundle);
+    CGH.single_task<TestKernel>([] {});
+  });
+  InOrderQueue.submit([&](sycl::handler &CGH) {
+    CGH.use_kernel_bundle(ExecBundle);
+    CGH.codeplay_host_task([=] {});
+  });
+  InOrderQueue.wait();
```

Comment:
Won't it trigger calls to `piEventsWait` for all tasks submitted to the queue?
Does this test fail without the fix?

Reply:
At least for OpenCL we now call `piQueueFinish(...)`, se we do get `piEventsWait(...)` at least once (before host task is enqueued).

The test fails without the fix.

Reply:
Ok, probably it would be more "robust" if we call `event::wait` for the second operation and check that `GEventsWaitCounter == 2`.

Reply:
Fixed